File: seven/SteeringController.cpp
/*
 * SteeringController.cpp
 *
 * ステアリング制御クラス（実装）
 * Follow the Gap + Pure Pursuit制御
 *
 * 設計思想:
 * - GapFinderが検出した目標方向と距離をPure Pursuit公式に適用
 * - 公式: steering = atan2(2 × L × sin(α), Ld)
 *   L: ホイールベース（mm）
 *   α: 目標点への角度（ラジアン）
 *   Ld: ルックアヘッド距離（mm）- 目標方向への推定距離
 */

#include "SteeringController.h"

#include <math.h>

#include "SensorReader.h"

// 度からラジアンへの変換定数
#ifndef DEG_TO_RAD
#define DEG_TO_RAD 0.017453292519943295f  // PI / 180.0
#endif

#ifndef RAD_TO_DEG
#define RAD_TO_DEG 57.29577951308232f  // 180.0 / PI
#endif

SteeringController::SteeringController() {}

void SteeringController::begin() {
    // Pure Pursuitはステートレスなので初期化処理なし
}

float SteeringController::calculate(const GapResult& gap, const SensorData* sensorData) {
    // GapFinderの出力を目標点の極座標として解釈
    // α (alpha): 目標点への角度
    // Ld: ルックアヘッド距離（目標点までの距離）

    float alpha_deg = gap.target_angle;

    // 壁の手前を目標点とする（壁からオフセット分手前）
    float Ld_mm = gap.target_distance - LOOKAHEAD_OFFSET_MM;

    // 安定性のためルックアヘッド距離をクランプ
    Ld_mm = constrain(Ld_mm, MIN_LOOKAHEAD_MM, MAX_LOOKAHEAD_MM);

    // 角度をラジアンに変換
    float alpha_rad = alpha_deg * DEG_TO_RAD;

    // Pure Pursuit公式: δ = atan2(2 × L × sin(α), Ld)
    float steering_rad = atan2(2.0f * WHEELBASE_MM * sin(alpha_rad), Ld_mm);

    // 度に変換
    float steering_deg = steering_rad * RAD_TO_DEG;

    // 最大操舵角でクランプ
    steering_deg = constrain(steering_deg, -MAX_STEERING_ANGLE, MAX_STEERING_ANGLE);

    return steering_deg;
}

void SteeringController::reset() {
    // Pure Pursuitはステートレスなのでリセット処理なし
}

File: src/python/1st - PID/SteeringController.py
# steering_controller.py
import math
import Config # <--- Importamos la configuración centralizada

class SteeringController:
    def __init__(self):
        # Pure Pursuit no tiene estado interno (stateless); el constructor está vacío.
        pass

    def begin(self):
        # Incluido por simetría con los otros módulos del sistema de Arduino
        pass

    def calculate(self, gap, sensor_data):
        """
        Calcula el ángulo de dirección (grados) usando Pure Pursuit.
        gap: Instancia de GapResult (contiene target_angle)
        sensor_data: Lista de objetos SensorData del SensorReader
        """
        alpha_deg = gap.target_angle

        # Restamos el offset a la distancia del sensor frontal para obtener Ld.
        # Equivalente al operador ternario ( ? : ) de C++
        front_sensor = sensor_data[Config.FRONT_SENSOR_INDEX]
        
        if front_sensor.valid:
            Ld_mm = front_sensor.distance - Config.LOOKAHEAD_OFFSET_MM
        else:
            Ld_mm = 1000.0  # Fallback de seguridad si el sensor frontal no es válido

        # Evitar división por cero o valores geométricos extremos/absurdos
        if Ld_mm < 50.0:
            Ld_mm = 50.0

        alpha_rad = alpha_deg * Config.DEG_TO_RAD

        # Fórmula matemática de Pure Pursuit: delta = atan2(2 * L * sin(alpha), Ld)
        steering_rad = math.atan2(
            2.0 * Config.WHEELBASE_MM * math.sin(alpha_rad), 
            Ld_mm
        )

        steering_deg = steering_rad * Config.RAD_TO_DEG

        # Clamp al límite mecánico del servo usando max() y min() (Equivalente al constrain)
       # 1. Aplicamos el límite (clamp) al desvío relativo usando tus constantes de Config
        steering_deg = max(-Config.MAX_STEERING_ANGLE, min(Config.MAX_STEERING_ANGLE, steering_deg))

        # 2. Convertimos el desvío relativo en un ángulo absoluto sumando tu centro (60)
        angulo_final_servo = Config.SERVO_CENTER_DEG + steering_deg

        # 3. Nos aseguramos de no sobrepasar los límites físicos reales configurados para tu servo
        # (Usa SERVO_MIN_DEG y SERVO_MAX_DEG de tu Config.py para proteger el mecanismo)
        angulo_final_servo = max(Config.SERVO_RIGHT_MAX_DEG, min(Config.SERVO_LEFT_MAX_DEG, angulo_final_servo))

        return angulo_final_servo